crates/ark/src/analysis/parse_boundaries.rs

```diff
@@ -0,0 +1,375 @@
+use harp::vector::Vector;
+use harp::ParseResult;
+use harp::RObject;
+
+/// Boundaries are ranges over lines of text.
+#[derive(Debug)]
+pub struct ParseBoundaries {
+    pub complete: Vec<std::ops::Range<usize>>,
+    pub incomplete: Option<std::ops::Range<usize>>,
+    pub error: Option<std::ops::Range<usize>>,
+}
+
+/// Parse boundaries of R inputs
+///
+/// Takes a string of R code and detects which lines parse as complete,
+/// incomplete, and error inputs.
+///
+/// Invariants:
+/// - There is always at least one range as the empty string is a complete input.
+/// - The ranges are sorted and non-overlapping.
+/// - Inputs are classified in complete, incomplete, and error sections. The
+///   sections are sorted in this order (there cannot be complete inputs after
+///   an incomplete or error one, there cannot be an incomplete input after
+///   an error one, and error inputs are always trailing).
+/// - There is only one incomplete and one error input in a set of inputs.
+pub fn parse_boundaries(text: &str) -> anyhow::Result<ParseBoundaries> {
+    let mut lines: Vec<&str> = text.lines().collect();
+
+    // Rectify for `lines()` ignoring trailing empty lines
+    match text.chars().last() {
+        Some(last) if last == '\n' => lines.push(""),
+        None => lines.push(""),
+        _ => {},
+    }
+
+    // Create a duplicate vector of lines on the R side too so we don't have to
+    // reallocate memory each time we parse a new subset of lines
+    let lines_r = CharacterVector::create(lines.iter());
+    let n_lines = lines.len();
+
+    let mut complete: Vec<std::ops::Range<usize>> = vec![];
+    let mut incomplete: Option<std::ops::Range<usize>> = None;
+    let mut error: Option<std::ops::Range<usize>> = None;
+
+    let mut incomplete_end: Option<usize> = None;
+    let mut error_end: Option<usize> = None;
+
+    let mut record_error = |start, error_end: &Option<usize>| {
+        if matches!(error, Some(_)) {
+            return;
+        }
+        let Some(end) = error_end else {
+            return;
+        };
+        error = Some(std::ops::Range { start, end: *end });
+    };
+
+    let mut record_incomplete = |start, incomplete_end: &Option<usize>| {
+        if matches!(incomplete, Some(_)) {
+            return;
+        }
+        let Some(end) = incomplete_end else {
+            return;
+        };
+        incomplete = Some(std::ops::Range { start, end: *end });
+    };
+
+    for current_line in (0..n_lines).rev() {
+        let mut record_complete = |exprs: RObject| -> anyhow::Result<()> {
+            let srcrefs = exprs.srcrefs()?;
+            let mut ranges: Vec<std::ops::Range<usize>> =
+                srcrefs.into_iter().map(|srcref| srcref.line).collect();
+
+            complete.append(&mut ranges);
+            Ok(())
+        };
+
+        // Grab all code up to current line
+        let subset = &lines_r.slice()[..current_line + 1];
+        let subset = CharacterVector::try_from(subset)?;
```

Comment:
I dislike that you have to do a (potentially large) redundant allocation on the 1st iteration, which most of the time is all you need to do (i.e. no incomplete and no error)

Reply:
Since you only reference `&subset` below, it does seem like you could add a special case for the first iteration that remapped `lines_r` directly as `subset` so you don't need that first allocation

I know that is uglier but it might be worth it

Reply:
How about this, with a `let mut first = true;` outside the loop?

```rust
        // Parse within source file to get source references.
        // Avoid allocation on the first iteration, since often there are no issues.
        let srcfile = if first {
            first = false;
            harp::srcref::SrcFile::try_from(&lines_r)?
        } else {
            // Grab all code up to current line
            let subset = &lines_r.slice()[..current_line + 1];
            let subset = CharacterVector::try_from(subset)?;
            harp::srcref::SrcFile::try_from(&subset)?
        };
```

Reply:
With the proposed approach we allocate one big string that is then split by R into a vector of lines (see R implementation of `srcfilecopy()`).

We could avoid the unnecessary splicing though. I made it:

```rust
        // Grab all code up to current line. We don't slice the vector in the
        // first iteration as it's not needed.
        let subset = if current_line == n_lines - 1 {
            CharacterVector::try_from(&lines_r.slice()[..=current_line])?
        } else {
            lines_r.clone()
        };
```

To support this, `CharacterVector` is now clonable (it does not deep-copy since none of our wrappers make owning guarantees for the wrapped data).